jupyter_server/base/websocket.py

```diff
@@ -82,6 +82,27 @@ def check_origin(self, origin: Optional[str] = None) -> bool:
     def clear_cookie(self, *args, **kwargs):
         """meaningless for websockets"""
 
+    @no_type_check
+    def _maybe_auth(self):
+        """Verify authentication if required"""
+        if not self.settings.get("allow_unauthenticated_access", False):
+            if not self.request.method:
+                raise web.HTTPError(403)
+            method = getattr(self, self.request.method.lower())
+            if not getattr(method, "__allow_unauthenticated", False):
+                # rather than re-using `web.authenticated` which also redirects
+                # to login page on GET, just raise 403 if user is not known
+                user = self.current_user
+                if user is None:
+                    self.log.warning("Couldn't authenticate WebSocket connection")
+                    raise web.HTTPError(403)
+
+    @no_type_check
+    def prepare(self, *args, **kwargs):
+        """Handle a get request."""
+        self._maybe_auth()
```

Comment:
`._maybe_auth` accesses `.current_user`, but `.prepare()` below is where it is set. The order needs to be reversed, but I assume there's a reason this is first. I think the override may need to be done a different way.

Reply:
Actually, why is IdentityProvider setting the current user in `JupyterHandler` and not in `AuthenticatedHandler` in the first place? I do not see rationale for it in #671

Reply:
Ah, I see the user from identity provider may require `await`ing. I guess I can just pass an argument to `JupyterHandler.prepare()` instructing it to not redirect to login page.

Reply:
JupyterHandler is responsible for `.current_user` being set, since it may be async, it cannot be done in a property. It would be appropriate to override the `.current_user` getter, which should make it very clear it can never be accessed before `.prepare()` is called:

```python
@property
def current_user(self):
    if not hasattr(self, "_jupyter_current_user"):
        raise RuntimeError(".current_user accessed before being populated in JupyterHandler.prepare()")
    return self. _jupyter_current_user

@current_user.setter
def current_user(self, user):
    self._jupyter_current_user = user
```

Any time that error is hit is necessarily code that is bypassing Jupyter Server authentication.

Reply:
Done in 5e7615d61d52210f497dd3d12ef7e8729599954a (added test) and 4265f4e6df6dcf60164a6bfcf158cdb03910e819 (fixed it).

Would you like me to add the `RuntimeError` on accessing `current_user` too early (from the last comment) in this PR, or would it be better to ensure that this PR does not introduce potentially breaking changes and can be adopted in existing deployments easily?

Reply:
Looking at the code, I guess we [already have this RuntimeError in get_current_user](https://github.com/jupyter-server/jupyter_server/blob/b3caa3c2f60412138f430daec1da7af235c8d26b/jupyter_server/base/handlers.py#L160-L176). So maybe it's just the message that wants clarifying?

Because it's already there, I don't think moving the RuntimeError to the `.current_user` getter can break anything. But at the same time, since folks _may_ call `get_current_user()` _or_ access the (correct) `.current_user`, then I guess `.get_current_user` covers both cases. So maybe it only makes sense to put it in `.current_user` if it helps to have two distinct, clear error messages.